src/promnesia/sources/reddit.py

```diff
@@ -3,40 +3,153 @@
+
+
+    def _from_comment(self, i: 'Comment') -> Results:
+        locator = Loc.make(
+            title='Reddit comment',
+            href=i.url,
+        )
+        yield from self._from_common(i, locator=locator)
+
+
+    def _from_submission(self, i: 'Submission') -> Results:
+        locator = Loc.make(
+            title=f'Reddit submission: {i.title}',
+            href=i.url,
+        )
+        yield from self._from_common(i, locator=locator)
+
+
+    def _from_upvote(self, i: 'Upvote') -> Results:
+        locator = Loc.make(
+            title=f'Reddit upvote',
+            href=i.url,
+        )
+        yield from self._from_common(i, locator=locator)
+
+
+    def _from_save(self, i: 'Save') -> Results:
+        locator = Loc.make(
+            title='Reddit save',
+            href=i.url,
+        )
+        yield from self._from_common(i, locator=locator)
+
+
+    # to allow for possible subclassing by the user?
+    def _render_body(self, text: str) -> str:
+        if self.render_markdown:
+            return self._parser_cls(text)._doc_ashtml()
+        else:
+            return text
+
+
+    def _from_common(self, i: 'RedditThing', locator: Loc) -> Results:
+        urls = [i.url]
+        # TODO this should belong to HPI.. fix permalink handling I guess
+        # ok, it's not present for all of them..
+        lurl = i.raw.get('link_url')
+        if lurl is not None:
+            urls.append(lurl)
+        lurl = i.raw.get('url')
+        if lurl is not None:
+            urls.append(lurl)
+
+        context = self._render_body(i.text)
+
+        emitted: Set[str] = set()
+
+        for url in chain(urls, extract_urls(i.text)):
+            if url in emitted:
+                continue
+            yield Visit(
+                url=url,
+                dt=i.created,
+                context=context,
+                locator=locator,
+            )
+            emitted.add(url)
+
+        # extract from markdown links like [link text](https://...)
+        # incase URLExtract missed any
+        #
+        # this should try to do this, even if the user didn't enable
+        # the render_markdown flag, as it may catch extra links that URLExtract didnt
+        # would still require mistletoe to be installed, but
+        # the user may already have it installed for the auto/markdown modules
+        if hasattr(self, '_link_extractor'):
+            for res in self._link_extractor(i.text):
+                if isinstance(res, Exception):
+                    yield res
+                    continue
+                if res.url in emitted:
+                    continue
+                yield Visit(
+                    url=res.url,
+                    dt=i.created,
+                    context=context,
+                    locator=locator,
+                )
+                emitted.add(res.url)
+
+
+def index(*, render_markdown: bool = False, renderer = RedditRenderer) -> Results:
```

Comment:
hmm, I usually like to keep `index` on the very top (so it's easier to follow what happens on the top level), but here it's got a default parameter, so needs to go after `RedditRenderer`... I guess it's ok

Reply:
Can make it optional and put it after the index function instead